_drafts/bash_completion.md

```diff
@@ -0,0 +1,179 @@
+
+ 1. Going deeper: it'd be nice if I could keep autocompleting deeper into the directory, so `cdx ydir/zdi`
+    could autocomplete to `~/x/ydir/zdir`. This would involve making the `opts=` line more intelligent.
+
+ 2. Fuzzy searching: I am great at typos. It'd be nice if I could replace `compgen` with something more
+    intelligent that will ignore small typos.
+
+ 3. More generic: Being able to match `cda` or `cdb` to switch to subdirectories of `~/apsis` or `~/bread`
+    would be nice. Writing a script to handle the unique cases when bash fires up would be easily doable.
+    It would, however, require separate autocomplete functions for each named function becuase `complete`
+    only accepts literal command names.
+
+## Troubleshooting/Further Reading
+
+If the above doesn't work it's possible your `progcomp` bash option isn't set. You can check with:
+
+`echo $BASHOPTS | grep -q progcomp && echo "YES" || echo "NO"`
+
+By default it's set, but if for whatever reason it's not, and you didn't knowingly do it yourself, you can set it
+with `shopt -s progcomp`. That can be added to your `.bashrc`.
+
+If that's not the problem, you will have to debug this the old fashioned way. Use `echo "$VARNAME"` to see the values
+of variables. Finally you may want to read the docs.
+
+I didn't initially suggest reading the docs because, unfortunately, the documentation for anything contained in the
+bash man pages is terribly undiscoverable, difficult to search, and not always easy to read [^3]. All of the above
+is in the bash man pages. Here's how you can find the relevant docs:
+
+ * `man bash` to open the documentation for bash itself.
+ * Press `/` to start searching.
+ * Once you've entered your search term hit "Enter" and then use `n` and `N` to search forward and backwards
+   (respectively) for the content you want.
+ * You'll want the following search terms:
+   * "Programming Completion" for an overview of completion in bash.
+   * "compgen" for the function that we use to match input with the search term.
+   * "complete" for the function used to tell bash to use programmable completion for a given command.
+
+## In Closing
+
+Bash Programmable Autocompletion is powerful, but intimidating at first. Hopefully this post helps you get started.
+If you build something awesome write it up and ping us on twitter! We're [@ApsisLabs](https://twitter.com/apsislabs).
+
+
+[^1]: Couldn't use an alias, because aliases can't interpolate arguments, they're strictly dumb text replacements , and they always include a space at the end.
+
+[^2]: [xkcd is relevant here](https://xkcd.com/1205/)
+
+[^3]: The bash man pages use tons of bolded references to dozens of other locations in a what is a huge doc with no links. This is becuase it's a man page, and man pages, as useful as they are, are dinosaurs [from a prehistoric time before hypertext was invented](https://unix.stackexchange.com/a/18161/34182).
```

Comment:
"dozens of other locations in a what is a huge doc with no links." You've got a few too many articles/prepositions in here.